tests/baselines/reference/definiteAssignmentOfDestructuredVariable.types

```diff
@@ -18,20 +18,20 @@ class C<T extends Options> {
 >method : () => void
 
         let { a, b } = this.foo;
->a : T["a"]
+>a : T["a"] | undefined
 >b : T["b"]
 >this.foo : { [P in keyof T]: T[P]; }
 >this : this
 >foo : { [P in keyof T]: T[P]; }
 
         !(a && b);
->!(a && b) : false

```

Comment:
so.... this was just... _wrong_ before? `a` was `T["a"]` which _should_ have been constrained to `number | object | undefined`, which means `a` can be falsey, which means the new result is correct.

Do we fail to include `undefined` constraints in our analysis somewhere in control flow?

Reply:
I believe it's the old issue of not having a good representation for the falsy parts of a higher-order type. For example:

```ts
interface Options {
    a: unknown;
    b: unknown;
}

function foo<T extends Options>(a: T['a'], b: T['b']) {
    const x = a && b;  // Type T['b']
}
```

The type of `x` really should be `T['b']` unioned with the falsy parts of `T['a']`, but even with negated types this would be a pretty ugly type. IIRC we previously decided to leave this as a known design limitation.

Reply:
Aw, yeah :(